EKF/control.cpp

```diff
@@ -1562,6 +1543,45 @@ void Ekf::controlMagFusion()
 	}
 }
 
+void Ekf::checkMagInhibition()
+{
+	_mag_use_inhibit = shouldInhibitMag();
+	if (!_mag_use_inhibit) { 
+		_mag_use_not_inhibit_us = _imu_sample_delayed.time_us;
+	}
+
+	// If magnetometer use has been inhibited continuously then a yaw reset is required for a valid heading
+	if (uint32_t(_imu_sample_delayed.time_us - _mag_use_not_inhibit_us) > (uint32_t)5e6) {
+		_mag_inhibit_yaw_reset_req = true;
+	}
+}
+
+bool Ekf::shouldInhibitMag() const
+{
+	// If the user has selected auto protection against indoor magnetic field errors, only use the magnetometer
+	// if a yaw angle relative to true North is required for navigation. If no GPS or other earth frame aiding
+	// is available, assume that we are operating indoors and the magnetometer should not be used.
+	// Also inhibit mag fusion when a strong magnetic field interference is detected
+	const bool user_selected = (_params.mag_fusion_type == MAG_FUSE_TYPE_INDOOR);
+	const bool not_using_gps = !(_params.fusion_mode & MASK_USE_GPS) || !_control_status.flags.gps;
+	const bool not_using_evpos = !(_params.fusion_mode & MASK_USE_EVPOS) || !_control_status.flags.ev_pos;
+	const bool not_using_evvel = !(_params.fusion_mode & MASK_USE_EVVEL) || !_control_status.flags.ev_vel;
+	const bool not_selected_evyaw =  !(_params.fusion_mode & MASK_USE_EVYAW);
+	const bool heading_not_required_for_navigation = not_using_gps &&
+							 not_using_evpos &&
+							 not_using_evvel &&
+							 not_selected_evyaw;
+
+	return (user_selected && heading_not_required_for_navigation)
+	       || isStrongMagneticDisturbance();
+}
+
+bool Ekf::isStrongMagneticDisturbance() const
+{
+	static constexpr float average_earth_mag_field_strength = 0.45f; // Gauss
+	return _mag_sample_delayed.mag.length() > 3.f * average_earth_mag_field_strength;
```

Comment:
It would probably also be reasonable to check for not enough field strength because the direction does not make sense anymore in that case.